client/verta/verta/runtime.py

```
@@ -0,0 +1,211 @@
+    Sets the thread-local logs, overwriting any existing values.
+    """
+    _THREAD.logs = logs
+    return _THREAD.logs
+
+
+def _init_validate_flag() -> None:
+    """
+    Initialize our thread-local variable for the validation flag.
+    """
+    if not hasattr(_THREAD, 'validate'):
+        _THREAD.validate = False
+
+
+def _get_validate_flag() -> bool:
+    """
+    Return the current thread-local variable for validate or initialize a
+    new boolean.
+    """
+    _init_validate_flag()
+    return _THREAD.validate
+
+
+def _set_validate_flag(flag: bool) -> bool:
+    """
+    Sets the thread-local validate flag boolean value.
+    """
+    _THREAD.validate = flag
+    return _THREAD.validate
+
+
+def _validate_json(value: Any) -> str:
+    """
+    Attempt to serialize the provided value to JSON.
+    Raise an error on failure.
+    """
+    try:
+        return json.dumps(value)
+    except json.JSONDecodeError as json_err:
+        print("provided data cannot be serialized into JSON for logging.")
+        raise json_err
+    except TypeError as type_err:
+        print("provided data type cannot be converted to JSON for logging.")
+        raise type_err
+
+
+def log(key: str, value: Any) -> None:
+    """
+    Updates current logging context dict with provided key and value.
+    For use within a model's :meth:`~verta.registry.VertaModelBase.predict`
+    method to collect logging context.
+
+    Parameters
+    ----------
+    key : str
+        String value to use as data label (key) in logging context.
+    value : Any
+        Any JSON serializable value you wish to include in the context logs.
+
+    Returns
+    -------
+    None
+
+    Raises
+    -------
+    TypeError
+        If `validate` was set to ``True`` on the active :class:`context`, and
+        `value` is not a JSON-serializable type.
+    JSONDecodeError
+        If `validate` was set to ``True`` on the active :class:`context`, and
+        `value` is not JSON-serializable.
```

I actually wouldn't expect to encounter `JSONDecodeError`s (https://github.com/VertaAI/modeldb/pull/3500#discussion_r1081979078) 🤔 Did you see any while testing/iterating?

---

I've been considering dropping this after trying to write a test for it.  It turns out it's really hard to trigger that error from `json.dumps()`.  I think you would have to read in json data from a file with malformed JSON or corrupted somehow.  It's technically possible, but I don't think we need to catch it.  It they manage to pull it off somehow, we can let the error bubble up naturally and feel just fine about it :)

---

Well, that is sort of what I'm asking about. You mention that "it's technically possible", but I don't actually think that it is. `dumps()` _encodes_ JSON, so I'm having trouble imaging how we'd encounter a decode error, and `json`'s documentation only mentions `JSONDecodeError` in its load methods.

So I was curious what motivated its inclusion here in the first place. Did you truly encounter any `JSONDecodeError` while working on this PR?

---

I think I just did some digging on what other sorts of exceptions could be encountered and stumbled across examples like [this](https://stackoverflow.com/questions/68000458/json-dump-lead-to-json-decoder-jsondecodeerror-when-json-loadf) and [this](https://stackoverflow.com/questions/55093573/json-loadsstring-throws-json-decoder-jsondecodeerror-when-trying-to-convert-st).  In any case, I don't think catching such a fringe exception makes sense.  I've already dropped all the references to it.